Input and output guardrails (#2571)
langchain4j/langchain4j: docs/docs/tutorials/guardrails.md

---
sidebar_position: 12
toc_max_heading_level: 5
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import ThemedImage from '@theme/ThemedImage';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Guardrails

Guardrails are mechanisms that let you validate the input and output of the LLM to ensure it meets your expectations. You can do some of the following things with guardrails:
- Verify the user input is not out of scope
- Ensure the input meets some criteria before calling the LLM (i.e. guard against a [prompt injection attack](https://genai.owasp.org/llmrisk/llm01-prompt-injection/))

> - Ensure the output format is correct (i.e. it is a JSON document with the correct schema)

Reviewer: nitpick: IMO one of the main use case for using output guardrails is making sure that the LLM output is coherent with the enterprise business rules and constraint. E.g.: if this is a chatbot of company X, I don't want in any case that the response could contain any reference to my competitor Y. Probably it could make sense to also mention this?

Author: Done!

- Detect hallucinations

Those are just examples. You can do many other things with guardrails.

:::note
Guardrails are only available when using [AI Services](/tutorials/ai-services). They are a higher-level construct that can not be applied to a `ChatModel` or `StreamingChatModel`.
:::

<ThemedImage
  alt="Guardrails"
  sources={{
    light: useBaseUrl('/img/guardrails-light-bg.png'),
    dark: useBaseUrl('/img/guardrails-dark-bg.png'),
  }}
/>;

> The implementation was originally done in the [Quarkus LangChain4j extension](https://docs.quarkiverse.io/quarkus-langchain4j/dev/) and was backported here. 

Reviewer: Does it make sense to mention this in langchain4j docs?

Author: I don't have a strong opinion either way. The Spring AI folks are touting themselves all over the place, so I thought this was a good chance to get some press :)

But I'm fine to get rid of it if you think it should not be there.

Reviewer 2: I would keep it, need to give credits

---
sidebar_position: 12
toc_max_heading_level: 5
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import ThemedImage from '@theme/ThemedImage';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Guardrails

Guardrails are mechanisms that let you validate the input and output of the LLM to ensure it meets your expectations. You can do some of the following things with guardrails:
- Verify the user input is not out of scope
- Ensure the input meets some criteria before calling the LLM (i.e. guard against a [prompt injection attack](https://genai.owasp.org/llmrisk/llm01-prompt-injection/))
- Ensure the output format is correct (i.e. it is a JSON document with the correct schema)
- Detect hallucinations

Those are just examples. You can do many other things with guardrails.

:::note

> Guardrails are only available when using [AI Services](/tutorials/ai-services). They are a higher-level construct that can not be applied to a `ChatLanguageModel` or `StreamingChatLanguageModel`.

Reviewer 2: ```suggestion
Guardrails are only available when using [AI Services](/tutorials/ai-services). They are a higher-level construct that can not be applied to a `ChatModel` or `StreamingChatModel`.
```

Also keep in mind that guardrails can themselves call other services or even invoke other LLM interactions. If these kinds of guardrails have an execution penalty or monetary cost associated with them, make sure you take that into account. You might want to put more expensive guardrails towards the end of the chain.

:::note
The term _expensive_ can mean that something takes some time to execute or has a monetary value associated with it.
:::

## Input Guardrails

Input guardrails are functions invoked before the LLM is called. Failing an input guardrail prevents the LLM from being called. Input guardrails are the last step prior to calling the LLM. They are invoked _after_ any [RAG](/tutorials/rag) operations have happened.

### Implementing Input Guardrails

Input guardrails are implemented by implementing the [`InputGuardrail`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/InputGuardrail.java) interface. The `InputGuardrail` interface has two variants of the `validate` method, at least one of which needs to be implemented:

```java
InputGuardrailResult validate(UserMessage userMessage);

> InputGuardrailResult validate(InputGuardrailParams params);

Reviewer 2: ```suggestion
InputGuardrailResult validate(InputGuardrailRequest params);
```
nit: do you mind renaming it into `InputGuardrailRequest` (and `GuardrailParams` to `GuardrailRequest`) to keep it in sync with `XyzRequest`/`XyzResult` naming pattern?

Author: Sure I will take care of this

Author: Done

### Declaring Input Guardrails

There are several ways to declare input guardrails, listed here in order of precedence:
1. [`@InputGuardrails` annotations](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/InputGuardrails.java) placed on an individual [AI Service](/tutorials/ai-services) method.
2. [`@InputGuardrails` annotation](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/InputGuardrails.java) placed on an [AI Service](/tutorials/ai-services) class.
3. [`InputGuardrail`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/InputGuardrail.java) implementation class names or instances set directly on the [`AiServices`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/AiServices.java) builder.
Regardless of how they are declared, input guardrails are always executed in the order they appear in the list.

#### Annotation on individual AI Service methods

> [`@InputGuardrails` annotations](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/InputGuardrails.java) placed on an individual [AI Service](/tutorials/ai-services) methods have the highest precendence, meaning if it is declared in any other ways, the one declared on the method will be the one used.

Reviewer 2: Shouldn't the guardrail specified in a programmatic way (via `AiServices`) have the higher precedence than one specified in a declarative way via annotation?

Author: Hmmmm interesting. Honestly I don't have strong feelings either way. I guess you are right that code should always override annotations?

If we agree that this is the better approach I will go ahead and make the necessary changes.

Reviewer 2: After a second thought, I would probably even prohibit mixing programmatic and annotation approaches, as it might lead to confusion.

Especially considering that there can be multiple guardrails. Should guardrails with higher precedence **_override_** those with lower precedence or **_append_** to them? Hovewer the same concern applies to annotation on class and annotation on method. WDYT? (Just thinking out loud here)

Author: > After a second thought, I would probably even prohibit mixing programmatic and annotation approaches, as it might lead to confusion.

I actually think its a good idea to allow all of them. As I've thought about it more I think your original comment is correct - specifying guardrails programmatically should have highest precendence. 

> Especially considering that there can be multiple guardrails. Should guardrails with higher precedence override those with lower precedence or append to them? Hovewer the same concern applies to annotation on class and annotation on method. WDYT? (Just thinking out loud here)

I didn't get that granular, nor do I think we should. If you have the annotation on a method, it overrides the annotation at the class level, period. It doesn't matter what values are in the annotations. The precendence is set by the presence of the annotation itself. This means we don't (& shouldn't) look at the individual lists of guardrails).

Reviewer 2: I do not have a strong preference, so let's go with what you propose

Author: I've made this change

```java
public interface Assistant {
    @InputGuardrails({ FirstInputGuardrail.class, SecondInputGuardrail.class })
    String chat(String question);
    
    String doSomethingElse(String question);
}

> var assistant = AiServices.create(Assistant.class, chatLanguageModel);

Reviewer 2: ```suggestion
var assistant = AiServices.create(Assistant.class, chatModel);
```

Could you please make sure there is no reference left to "chatLanguageModel" (please use "chatModel" instead) and "streamingChatLanguageModel" (please use "streamingChatModel" instead)?

Author: Done

import dev.langchain4j.data.message.UserMessage;
import dev.langchain4j.guardrail.GuardrailResult.Result;

class Tests { 
    MyInputGuardrail inputGuardrail = new MyInputGuardrail();
    
    @Test 
    void test() {
        var userMessage = UserMessage.from("Some user message");
        var result = inputGuardrail.validate(userMessage);
        
        // These are just some examples of what you can do
        assertThat(result)
                .isSuccessful()
                .hasResult(Result.FATAL)
                .hasFailures()
                .hasSingleFailureWithMessage("Prompt injection detected")
                .assertSingleFailureSatisfied(failure -> assertThat(failure)...)
                .withFailures().....
    }
}
```

:::info
See the [`GuardrailAssertions`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-test/src/main/java/dev/langchain4j/test/guardrail/GuardrailAssertions.java) and [`InputGuardrailResultAssert`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-test/src/main/java/dev/langchain4j/test/guardrail/InputGuardrailResultAssert.java) classes for more details.
:::

## Output Guardrails

> Output guardrails are functions executed after the LLM has produced its output. Failing an output guardrail allows for more advanced scenarios, such as [retrying](#retry) or [reprompting](#reprompt), to help improve the response. They are invoked _after_ all other operations, including function/tool calls, have happened.

Reviewer 2: Is there a plan to have dedicated tool guardrails in the future? Otherwise it would make sense to invoke guardrails before tools are called?

Author: I think so yes. These guardrails are general guardrails for an LLM interaction. I think getting more specific for something like tool invocations would be good to add in the future, but isn't necessarily related to this.

Author: > I think getting more specific for something like tool invocations would be good to add in the future, but isn't necessarily related to this.

Thinking about this, maybe instead of `InputGuardrail`/`OutputGuardrail` we rename everything to `LLMInteractionInputGuardrail`/`LLMInteractionOutputGuardrail`.

Then the annotations can remain, and maybe `InputGuardrail`/`OutputGuardrail` can remain as a base interface that could be extended in the future (i.e. `ToolInvocationInputGuardrail`/`ToolInvocationOutputGuardrail`). In which case everything we've built from the execution perspective could simply be re-used.

Just thinking out loud I suppose. I don't want to over-engineer now, but also knowing that we may want to do something like this in the future, trying to think of a way to make it more backwards-compatible now so that when we do it in the future it isn't a ton (if any?) breaking changes.

Reviewer 2: Correct me if I am wrong, but validating LLM-generated tool calls (names and arguments of functions) is probably one of the most interesting use cases where output guardrails shine?

As a user I would probably expect that output guardrails are called before any tool is called as tool call **_is_** an LLM output, although specific.

Ofcourse tool-specific guardrails can be added a bit later to make it easier to implement it for the user (e.g. provide parsed arguments instead of a raw JSON string), but for the time being there probably should be a more generic feature that can provide a flexibility to implement validation of tool call (even if not perfectly convenient).

I am not sure I fully understand the `LLMInteractionOutputGuardrail` suggestion, as in my understanding tool call fits into both "LLM interaction" and "output", so `LLMInteractionOutputGuardrail` should conver tool calls as well.

Do you see a big issue with calling output guardrail also before tools are called?

Author: So I know where you are going with your comments. The guardrails are intended to guard against an "LLM interaction", where an "LLM interaction" is defined as an operation giving input and getting output. They are intended to run before the initial LLM call is invoked and just before the result is returned _to the user_.

```java
var chatModel = ...

var response = chatModel.chat(request);
```

What happens _during_ the interaction doesn't matter from the guardrails perspective (i.e. multiple potential round-trips to the model, or erven other models, as a guardrail itself _could_ invoke another model, which could have its own set of guardrails.

I think if we want to have guardrails for tool invocations (which is a good idea), they would be slightly different. They would have different inputs/outputs, and they would need to be declared differently. You wouldn't necessarily declare them on the ai service (maybe you would, but I haven't yet thought through it).

My point is, these guardrails are intended to encapsulate an entire llm interaction, which may involve multiple llm interactions.

I think adding tool guardrails should be dealt with separately in a separate PR, as they are different.

> I am not sure I fully understand the `LLMInteractionOutputGuardrail` suggestion

I probably didn't describe it properly. My thought was maybe trying to add another layer in here to be able to distinguish between these kinds of guardrails & future tool invocation guardrails. I don't want to over-complicate or over-engineer yet though.

Reviewer 2: > I think if we want to have guardrails for tool invocations (which is a good idea), they would be slightly different. They would have different inputs/outputs, and they would need to be declared differently

Technically you can validate tools by implementing `OutputGuardrail.validate(AiMessage)` as tool invocation are inside the `AiMessage`. But I see your point and agree that tool guardrails should be specified separately

### Implementing Output Guardrails

> Similar to input guardrails, output guardrails are implemented by implementing the [`OutputGuardrail`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/OutputGuardrail.java) interface. The `OutputGuardrail` interface has two variants of the `validate` method, at least one of which needs to be implemented:

Reviewer 2: TODO (for myself): guardrail logic location: core/main

| **_fatal_**                | `fatal(String)` or `fatal(String, Throwable)`                       | The output is invalid and execution is halted with an `OutputGuardrailException` thrown to the caller.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |

> | **_fatal with retry_**     | `retry(String)` or `retry(String, Throwable)`                       | - Similar to **_fatal_** except the LLM is called again with the same prompt and chat history as the original call.<br/> - If the failure persists after a [configurable number of retries](#configuration) then execution is halted with an `OutputGuardrailException` thrown to the caller.<br/> - If the guardrail passes after a retry, the entire chain of guardrails are re-executed from the beginning.                                                                                                                                                                                                        |

Reviewer 2: Is there a limit to how many times LLM is called overall? Otherwise there is a risk of a deadlock: 
1. first guardrail passes and second one fails
2. re-execution happens
3. first guardrail fails
4. re-execution happens
5. goto 1

Author: hmmm interesting. Thats something I didn't think about, and I don't believe we have in Quarkus today (@geoand / @mariofusco / @jmartisk ?).

Maybe we apply the retry logic to the set of guardrails as wel as each individual guardrail?

Reviewer: No, this is something that we don't have in the Quarkus extension, and I agree that it would be very useful. My only doubt is: is this something that we want to configure on the guardrail itself or on the AI service using it? I'm asking also because in the second case we will also need to change both the API and the annotation used to configure a guardrail on a service.

Author: @dliubarskyi Is this something we want to add as part of this large PR? Or as an enhancement later on?

> My only doubt is: is this something that we want to configure on the guardrail itself or on the AI service using it? I'm asking also because in the second case we will also need to change both the API and the annotation used to configure a guardrail on a service.

This is one reason why I'd like to punt this to a future enhancement. Let's get this in there and then we can refine it.

> | **_fatal with reprompt_**  | `reprompt(String, String)` or `reprompt(String, Throwable, String)` | - Similar to **_fatal with retry_** except the LLM is called again with a new prompt supplied by the guardrail.<br/> - In this situation, the guardrail supplies an additional message to append to the previous user message, then sends a new request to the LLM with the new user message and original chat history.<br/> - If the failure persists after a [configurable number of retries](#configuration) then execution is halted with an `OutputGuardrailException` thrown to the caller.<br/> - If the guardrail passes after a reprompt, the entire chain of guardrails are re-executed from the beginning. |

Reviewer 2: In case of retry and reprompt, do we keep unsuccessful `AiMessage`s and retry `UserMessage`s in the `ChatMemory`? I would not add these to the `ChatMemory` by default and make it configurable later if users will want to keep them. WDYT?

Author: I believe we do keep them in the memory, because its feedback to the LLM of what its done, and more importantly, what it did wrong.

@geoand / @cescoffier / @jmartisk maybe you have more info as to why keeping these "bad" messages in memory is a good idea?

Reviewer 2: From what I've seen recently, LLM guides recommend telling LLM what to do rather than what not to do. Same applies for examples I guess?

Also, since it is cheaper (in terms of tokens, LLM latency and memory consumption) to not add "bad" `AiMessage`+`UserMessage` pairs to the `ChatMemory`, I would opt for it to be the default behaviour. If there will be a reason for users to do otherwise, we can always add a configuration option later to change this.

Author: @geoand / @cescoffier / @jmartisk / @mariofusco  maybe you have more info as to why keeping these "bad" messages in memory is a good idea?

I see and can understand @dliubarskyi 's points, but I also am not aware of all the reasons why Quarkus decided to go the other way.

Reviewer: I see pros and cons in both alternatives. I agree with @dliubarskyi that not adding bad messages to the memory will help to reduce the size of tokens and probably to keep the LLM more focused on the assigned task. On the other side a bad message could still be useful as a negative example (this mostly depends on the underlying model I guess) and not adding it will require some memory rewriting that it is always risky and should be avoided except when absolutely necessary.

Ideally I would make this configurable, but if I should choose between one of the 2 (at least as default behavior), I would prefer to keep everything in the memory, including bad messages.

Reviewer 2: > not adding it will require some memory rewriting that it is always risky and should be avoided except when absolutely necessary

There is no need to **_rewrite_** the memory, just not add the `AiMessage` that failed the guardrail and not add the retry/reprompt `UserMessage` to the `ChatMemory` in the first place. You can see it as "add only those messages to the memory which passed the validation and which user actually can access". In case of end user interacting with AI service, end user will not see the messages which were failed and will not see retry attempts, so keeping them in `ChatMemory` is against WYSIWYG I guess.

I do not know which approach is better in terms of the quality/success, as it would require extensive evaluation and depend on the model. But I surely know that not keeping "bad" messages in the memory as a default behaviour will be cheaper and faster for users. I usually optimize for cost in such cases

Author: I will go ahead and make this change. THanks @mariofusco & @dliubarskyi for the conversation!

Author: Done!

### Declaring Output Guardrails

There are several ways to declare output guardrails, listed here in order of precedence:
1. [`@OutputGuardrails` annotations](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/OutputGuardrails.java) placed on an individual [AI Service](/tutorials/ai-services) method.
2. [`@OutputGuardrails` annotation](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/OutputGuardrails.java) placed on an [AI Service](/tutorials/ai-services) class.

> 3. [`OutputGuardrail`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/OutputGuardrail.java) implementation class names or instances set directly on the [`AiServices`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/AiServices.java) builder.

Reviewer 2: same question about precedence (programamtic vs declarative)

Author: Same response [as above](https://github.com/langchain4j/langchain4j/pull/2571#discussion_r2093297938)

}

var assistant = AiServices.builder(Assistant.class)
    .chatLanguageModel(chatLanguageModel)
    .outputGuardrails(new FirstOutputGuardrail(), new SecondOutputGuardrail())
    .build();
```

In the first scenario, classes that implement `OutputGuardrail` are passed. New instances of these classes are created dynamically using reflection.

:::info
The way classes are converted to instances can be customized. For example, frameworks that use dependency injection (like [Quarkus](https://quarkus.io) or [Spring](https://spring.io)) can use [extension points](#extension-points) to provide instances based on how they manage class instances rather than creating new instances via reflection each time.
:::

#### Configuration

Output guardrails have the following additional configuration that can be supplied:

| Configuration | Description                                                                                                                                                |
|:--------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `maxRetries`  | - The maximum number of retries for an output guardrail when performing a retry or reprompt.<br/> - Defaults to `3`.<br/> - Set to `0` to disable retries. |

Reviewer 2: nit: I have recently changed the default of `maxRetries` to `2` everywhere (maximum 3 calls in total), perhaps it makes sense to have 2 here as well for consistency?

Author: Done!

Similar to the `OutputGuardrail` classes, there are several ways to declare output guardrail configuration, listed here in order of precedence:
1. [`@OutputGuardrailsConfig` annotation](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/OutputGuardrailsConfig.java) placed on an individual [AI Service](/tutorials/ai-services) method.
2. [`@OutputGuardrailsConfig` annotation](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/OutputGuardrailsConfig.java) placed on an [AI Service](/tutorials/ai-services) class.
3. [`OutputGuardrailsConfig`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/config/OutputGuardrailsConfig.java) instance set directly on the [`AiServices`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/AiServices.java) builder.

##### Annotation on individual AI Service methods

```java
public interface MethodLevelAssistant {
    @OutputGuardrails(
            value = { FirstOutputGuardrail.class, SecondOutputGuardrail.class },

>             config = @OutputGuardrailsConfig(maxRetries = 10)

Reviewer 2: nit: does it make sense to flatten the `config` and have `@OutputGuardrails(value = {...}, maxRetries = 10)` ?

Author: I thought about that, but then I was thinking that in the future maybe there might be more config options? I somewhat modeled this after OpenApi. Doing it this way just shields us in the future from a backwards-compatibility standpoint if we start to see the need for lots of additional config and we don't want to pollute the `@OutputGuardrails` annotation.

I don't have a strong preference either way though.

Reviewer 2: > Doing it this way just shields us in the future from a backwards-compatibility standpoint if we start to see the need for lots of additional config and we don't want to pollute the @OutputGuardrails annotation.

Could you please explain regarding backwards-compatibility, how doe it shield?

Author: Right now its pretty simple, but in the future if the number of parameters were to grow, you could potentially have an annotation (`@OutputGuardrails`) that had lots of attributes. By separating it out you are encapsulating any parameters that are `configuration-related` together into its own sub-category. Kinda like creating an object structure and keeping related attributes together in objects and then composing objects together.

But like I mentioned, I don't have a strong preference. This is just something I usually do by default as a best practice. If you'd prefer to flatten then thats fine with me too.

Just if you flatten now and decide later you want to categorize/group things, it would be a breaking change.

Reviewer 2: Do you anticipate lots of configuration parameters in the future? If yes, let's keep it as it is. If no, I would flatten as it seems too verbose

Author: Done!

```

##### Annotation on the AI Service class

```java
@OutputGuardrails(
        value = { FirstOutputGuardrail.class, SecondOutputGuardrail.class },
        config = @OutputGuardrailsConfig(maxRetries = 10)
)
public interface ClassLevelAssistant {
    String chat(String question);
}

var assistant = AiServices.create(ClassLevelAssistant.class, chatLanguageModel);
```

##### `AiServices` builder

```java
public interface Assistant {
    String chat(String message);
}

var outputGuardrailsConfig = OutputGuardrailsConfig.builder()
        .maxRetries(10)
        .build();

var assistant = AiServices.builder(Assistant.class)
        .chatLanguageModel(chatLanguageModel)
        .outputGuardrailsConfig(outputGuardrailsConfig)
        .outputGuardrailClasss(FirstOutputGuardrail.class, SecondOutputGuardrail.class)
        .build();
```

### Output Guardrails on Streaming Responses

Output guardrails can also work for operations with streaming responses:

```java
public interface StreamingAssistant {
    @OutputGuardrails({ FirstOutputGuardrail.class, SecondOutputGuardrail.class })
    TokenStream streamingChat(String message);
}
```

> In this scenario, the output guardrails will be executed once the entire stream is complete, or more specifically, when `TokenStream.onCompleteResponse` is called.

Reviewer 2: I am not sure how useful this is? Usually streaming is done when the user is awaiting the response ASAP. So calling guardrails when user already read the answer does not seem useful? Or do you have some good use case in mind?

Author: Honestly I had somewhat of a difficult time with the streaming aspect, because you can't apply a guardrail until you've read the entire response.

In Quarkus, we buffer the response and then apply the output guardrail to the buffer, but here in LC4j this isn't "true" streaming. Its not asynchronous - its just a synchronous callback. To _really_ implement streaming we'd need to implement Java `Flow`s/etc, and that wasn't something I wanted to take up as part of this :)

Reviewer 2: > In Quarkus, we buffer the response and then apply the output guardrail to the buffer

Does it mean that user does not see anything untill the LLM is done streaming?

Author: Thats a good question and probably needs some though. I think it depends on the user's implementation of the `TokenStream`. If they implement `onPartialResponse` then maybe the end user sees stuff as its happening?

I think maybe we (I) need to re-think how streaming should work, meaning if there are output guardrails declared do we buffer the entire response first, apply guardrails, then iterate through the response and return it? We'd essentially be converting from streaming -> synchronous.

Author: This is the approach I took. If output guardrails are declared for a particular method, in `AiServiceStreamingResponseHandler` I intercept `onPartialResponse` and instead buffer the response. Then after I've applied the guardrails and all is good I replay the buffer.

Thats the best/simplest thing I can think of without going to fully non-blocking.

> In the situation where a **_retry_** or **_reprompt_** in the chain eventually succeeds, then the entire chain is re-executed _synchronously_. Once the chain completes the result is passed into `TokenStream.onCompleteResponse`.

Reviewer 2: Could you please add a bit more details about the "_synchronously_" part? What exactly does it mean? What happens with the tokens that are streamed? What happens if tools are executed in-between?

Author: Done

|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | [`JsonExtractorOutputGuardrail`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/JsonExtractorOutputGuardrail.java) | An output guardrail that will check whether or not a response can be successfully deserialized from JSON to an object of a certain type.<br/> - Uses a [Jackson ObjectMapper](https://github.com/FasterXML/jackson-databind) to try and deserialize an object.<br/> - The LLM is reprompted if the response can't be deserialized into the expected object type.<br/> - Can be used as-is, or can be extended and customized (there are several `protected` methods that can be overridden to customize behavior). |

Reviewer 2: TODO (for myself): `JsonExtractorOutputGuardrail`'s name

Author: Yeah I wasn't crazy about the name either :) Maybe `MarshalsIntoJsonObjectOutputGuardrail`?

Reviewer 3: What about `OutputGuardrailJsonExtractor` ?

Reviewer 2: Maybe `ValidJsonOutputGuardrail` or something similar? Basically this guardrail checks that json is valid and can be deserialized into the target object

### Declaring Input Guardrails

There are several ways to declare input guardrails, listed here in order of precedence:
1. [`@InputGuardrails` annotations](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/InputGuardrails.java) placed on an individual [AI Service](/tutorials/ai-services) method.
2. [`@InputGuardrails` annotation](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/guardrail/InputGuardrails.java) placed on an [AI Service](/tutorials/ai-services) class.
3. [`InputGuardrail`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-core/src/main/java/dev/langchain4j/guardrail/InputGuardrail.java) implementation class names or instances set directly on the [`AiServices`](https://github.com/langchain4j/langchain4j/blob/main/langchain4j/src/main/java/dev/langchain4j/service/AiServices.java) builder.

> 4. 

Reviewer 4: ```suggestion
```
